### DISCUSSION TITLE
Layering system?

### DISCUSSION THREAD

**Discussion by James009D** [hlky/stable-diffusion-webui#925] [Feature suggestions]
Would it be possible to have a layers system where we could do have foreground, mid, and background objects which relate to one another and share the style? So we could say generate a landscape, one another layer generate a castle, and on another layer generate a crowd of people.

**Comment by codedealer** [comment]
Needs transparency masking to work but it's possible.

But realistically you need to look into SD plugins for image editors like Photoshop or Krita. It's easier to bring SD into Photoshop than Photoshop into SD.

**Comment by xaedes** [comment]
Yea, I had some free time this weekend and made a working prompt-based layering system. 
Basically you write a a multi-line prompt that looks like markdown, where each section declares one layer.
It is hierarchical, so each layer can have their own child layers.

Examples: https://imgur.com/a/eUxd5qn
![](https://i.imgur.com/L61w00Q.png)

Sections with "prompt" and child layers are img2img, without child layers they are txt2img.
Without "prompt" they are just images, useful for mask selection, image composition, etc.
Images can be initialized with "color", resized with "resize" and their position specified with "pos". 
Rotation and rotation center are "rotation" and "center".

It even has automatic mask selection based on https://huggingface.co/spaces/atsantiago/Monocular_Depth_Filter

![](https://i.imgur.com/8rMHWmZ.png)

I am currently working on some last things (especially needs some user documentation) and 
make a PR later this week. 

You are welcome to test it from my development branch:

https://github.com/xaedes/stable-diffusion-webui/tree/scn2img-mask-by-estimated-depth

Later I will add 3D-Scene Reconstruction based on the depth estimation, which gives us 3D-Rotation of images!
It won't work good everywhere, but good enough to show something from little bit different perspective.
